### DISCUSSION TITLE
Setting selected option in ComboBox

### DISCUSSION THREAD

**Discussion by matthewcrews** [fsprojects/Avalonia.FuncUI#334] [Q&A] [answered]
I am trying to create a ComboBox with one of the options already selected. My code:

```fsharp
ComboBox.create [
    ComboBox.selectedIndex 0
    ComboBox.viewItems [
        ComboBoxItem.create [
            ComboBoxItem.content "Priority"
            ComboBoxItem.onPointerPressed (fun e ->
                ())
        ]
        ComboBoxItem.create [
            ComboBoxItem.content "Proportional"
            ComboBoxItem.onPointerPressed (fun e ->
                ())
        ]
    ]
]
```

What I get:
![image](https://github.com/fsprojects/Avalonia.FuncUI/assets/1593215/bd12ae38-2aee-427a-85d5-811a441a0ad7)

What I expect:
![image](https://github.com/fsprojects/Avalonia.FuncUI/assets/1593215/447e0bba-83b5-4951-936e-43d4f51f2187)

**Comment by JordanMarr** [comment]
You have hard-coded it to select item 0.
Normally you would store the selected item in your `Model` and then bind that to `ComboBox.selectedItem`.

https://github.com/fsprojects/Avalonia.FuncUI/blob/d959cfe170477dda2f7719b5bb03539a99eac771/src/Avalonia.FuncUI.ControlCatalog/Avalonia.FuncUI.ControlCatalog/Views/Tabs/FlyoutDemo.fs#L61

**Reply by JaggerJo** [reply, marked as answer]
You need to change the order of attributes.

First specify the options, then specify the selected index.

**Reply by JordanMarr** [reply]
😮

**Reply by matthewcrews** [reply]
Changing the order of the attributes fixed it. Strange that this doesn't raise an error at runtime. That's what I would expect.

**Reply by JaggerJo** [reply]
It's just how it's implemented in Avalonia. FuncUI just calls property setters behind the covers.

It makes sense if you think about it generally. 

Assume you have only a 3 items, the last one is selected. Now the last element is removed. This will clear the selection.

Also note that 'SelectedIndex' is just one possible way of setting the selection.

**Reply by JordanMarr** [reply]
That makes sense that it's imperative and therefore the order matters.

The confusion likely stems from comparing the FuncUI DSL to declarative style XAML bindings where the order doesn't matter:
```xaml
        <ComboBox SelectedIndex="{Bindings SelectedIndex}" Items="{Bindings MyData}">
```

**Reply by edgarfgp** [reply]
This is one the reason why in Fabulous we have widgets constructor to initialise the controls so the order does not matter. In v1 we had similar problems 😓

**Reply by matthewcrews** [reply]
FYI, I love AvaloniaFuncUI. It's the most productive I've ever been with a UI library. These edge cases often send me into a confusion spiral though 🤣.

**Reply by JordanMarr** [reply]
That makes perfect sense! 
I was thinking there would need to be an intermediary phase where declarative settings could be listed and then sorted.

**Reply by JaggerJo** [reply]
I'm not a fan of doing something special regarding attribute order as it's per control / subclass of a control.

@edgarfgp how does fabulous determine the order? 🤔

Is it only done for certain attributes?

IMHO it's just something to be aware of. Even handy once you know things are applied in that order always.
